azurelinuxagent/common/protocol/imds.py

```diff
@@ -242,15 +243,46 @@ def __init__(self, version=APIVERSION):
             'User-Agent': restutil.HTTP_USER_AGENT_HEALTH,
             'Metadata': True,
         }
-        pass
+        self._regex_imds_ioerror = re.compile(r".*HTTP Failed. GET http://[^ ]+ -- IOError timed out -- [0-9]+ attempts made")
+        self.protocol_util = get_protocol_util()
 
-    @property
-    def compute_url(self):
-        return BASE_URI.format(IMDS_ENDPOINT, 'compute', self._api_version)
+    def _get_metadata_url(self, endpoint, resource_path):
+        return BASE_URI.format(endpoint, resource_path, self._api_version)
 
-    @property
-    def instance_url(self):
-        return BASE_URI.format(IMDS_ENDPOINT, '', self._api_version)
+    def _http_get(self, endpoint, resource_path, headers):
+        url = self._get_metadata_url(endpoint, resource_path)
+        return restutil.http_get(url, headers=headers, use_proxy=False)
+
+    def get_metadata(self, resource_path, is_health):
+        """
+        Get metadata from IMDS, falling back to Wireserver endpoint if necessary.
+
+        :param str resource_path: path of IMDS resource
+        :param bool is_health: True if for health/heartbeat, False otherwise
+        :return: Tuple<is_request_success:bool, response:str>
+            is_request_success: True when connection succeeds, False otherwise
+            response: response from IMDS on request success, failure message otherwise
+        """
+        headers = self._health_headers if is_health else self._headers
+        endpoint = IMDS_ENDPOINT
+        try:
+            resp = self._http_get(endpoint=endpoint, resource_path=resource_path, headers=headers)
+        except HttpError as e:
+            logger.warn("Unable to connect to primary IMDS endpoint {0}", endpoint)
```

Comment:
Once a minute based on MonitorHandler here: https://github.com/Azure/WALinuxAgent/blob/develop/azurelinuxagent/ga/monitor.py#L114

Reply:
let's make it periodic, see https://github.com/Azure/WALinuxAgent/blob/develop/azurelinuxagent/common/event.py#L293 for an example

Reply:
Done.

Reply:
There are log messages that can potentially be made periodic in monitor.py that logs with IMDS is healthy or not.

I think it would require a new feature of periodic logging that accounts for state change where IMDS health can change states (flip between healthy and unhealthy). We'd want to capture that state change in the logs, which should be rare making periodic logging meaningful and desirable most of the time.

Let me know your thoughts. Thanks!

Reply:
Yes, I have also seen sections of the code where periodic logging with state changes is needed. Currently the state flip and logging is handled explicitly in that code. Next time I touch it I'll try to refactor this functionality to the logger.